### PR TITLE
EVG-5663: avoid expensive update operation during task dispatching

### PR DESCRIPTION
**Pull request by tychoish**

### CODE REVIEW

**Comment by tychoish** [comment]
Ok, I've staged this so that the early stage should be safe to deploy and revert to, and then we can change the feature flag and use the new mechanism so that the revert is safe.

File: makefile [hunk 1]
```diff
@@ -68,6 +68,7 @@ lintArgs += --disable="megacheck" --enable="unused" --enable="gosimple"
 lintArgs += --exclude=".*([mM]ock.*ator|modadvapi32|osSUSE) is unused \((deadcode|unused|megacheck)\)$$"
 lintArgs += --exclude="error return value not checked \((defer .*|fmt.Fprint.*) \(errcheck\)$$"
 lintArgs += --exclude=".* \(SA5001\) \(megacheck\)$$"
+lintArgs += --exclude=".*file is not goimported"
```
[the remark this answers is not in the document]
Comment (tychoish) [reply makefile]: I think this specific tag is useless in the way that we're running metalinter.

File: model/task_queue.go [hunk 2]
```diff
@@ -302,20 +309,82 @@ func ClearTaskQueue(distro string) error {
...
+			"$unwind": bson.M{
+				"path": "$" + taskQueueQueueKey,
+				"preserveNullAndEmptyArrays": true,
+			},
+		},
+		{
+			"$match": bson.M{
+				"$or": []bson.M{
+					{ // queue is empty
+						taskQueueQueueKey: bson.M{"$exists": false},
```
[the remark this answers is not in the document]
Comment (tychoish) [reply model/task_queue.go]: not after a $unwind, it's just the value of that element.

**Comment by tychoish** [comment]
The aggregation should be isomorphic with the find. This is just the only way that I could find (on short notice) to project out elements of an embedded array...

**Comment by john-m-liu** [comment]
That seems unnecessary given that you aren't projecting out anything significant. This aggregation almost certainly performs worse than the previous query

**Comment by tychoish** [comment]
I think you're probably right, this gives us a quick (and safe to deploy) path toward being able to have a MUCH less expensive operation available for removing/dispatching items from the queue, and I think once we're safe from reverts on this code, we can remove some of the backwards compatibility stuff in this agg to make it faster.

**Pull request merged**